source/windows/bcrypt_rsa.c

```diff
@@ -0,0 +1,459 @@
+
+    BCRYPT_RSAKEY_BLOB key_blob;
+    AWS_ZERO_STRUCT(key_blob);
+    key_blob.Magic = BCRYPT_RSAFULLPRIVATE_MAGIC;
+    key_blob.BitLength = (ULONG)private_key_data.modulus.len * 8;
+    key_blob.cbPublicExp = (ULONG)private_key_data.publicExponent.len;
+    key_blob.cbModulus = (ULONG)private_key_data.modulus.len;
+    key_blob.cbPrime1 = (ULONG)private_key_data.prime1.len;
+    key_blob.cbPrime2 = (ULONG)private_key_data.prime2.len;
+
+    struct aws_byte_cursor header = aws_byte_cursor_from_array(&key_blob, sizeof(key_blob));
+    aws_byte_buf_append(&key_pair_impl->key_buf, &header);
+
+    LPCWSTR blob_type = BCRYPT_RSAFULLPRIVATE_BLOB;
+    ULONG flags = 0;
+
+    aws_byte_buf_append(&key_pair_impl->key_buf, &private_key_data.publicExponent);
+    aws_byte_buf_append(&key_pair_impl->key_buf, &private_key_data.modulus);
+    aws_byte_buf_append(&key_pair_impl->key_buf, &private_key_data.prime1);
+    aws_byte_buf_append(&key_pair_impl->key_buf, &private_key_data.prime2);
+    aws_byte_buf_append(&key_pair_impl->key_buf, &private_key_data.exponent1);
+    aws_byte_buf_append(&key_pair_impl->key_buf, &private_key_data.exponent2);
+    aws_byte_buf_append(&key_pair_impl->key_buf, &private_key_data.coefficient);
+    aws_byte_buf_append(&key_pair_impl->key_buf, &private_key_data.privateExponent);
+
+    NTSTATUS status = BCryptImportKeyPair(
+        s_rsa_alg,
+        NULL,
+        blob_type,
+        &key_pair_impl->key_handle,
+        key_pair_impl->key_buf.buffer,
+        (ULONG)key_pair_impl->key_buf.len,
+        flags);
+
+    if (!BCRYPT_SUCCESS(status)) {
+        aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        goto on_error;
+    }
+
+    key_pair_impl->base.vtable = &s_rsa_key_pair_vtable;
+    key_pair_impl->base.key_size_in_bits = private_key_data.modulus.len * 8;
+    key_pair_impl->base.good = true;
+
+    aws_der_decoder_destroy(decoder);
+
+    return &key_pair_impl->base;
+
+on_error:
+    aws_der_decoder_destroy(decoder);
+    aws_byte_buf_clean_up_secure(&key_pair_impl->base.priv);
+    aws_byte_buf_clean_up_secure(&key_pair_impl->base.pub);
+    s_rsa_destroy_key(&key_pair_impl->base);
+    return NULL;
+}
+
+struct aws_rsa_key_pair *aws_rsa_key_pair_new_from_public_key_pkcs1_impl(
+    struct aws_allocator *allocator,
+    struct aws_byte_cursor key) {
+
+    aws_thread_call_once(&s_rsa_thread_once, s_load_alg_handle, NULL);
+    struct bcrypt_rsa_key_pair *key_pair_impl = aws_mem_calloc(allocator, 1, sizeof(struct bcrypt_rsa_key_pair));
+
+    aws_ref_count_init(&key_pair_impl->base.ref_count, &key_pair_impl->base, aws_rsa_key_pair_destroy);
+    key_pair_impl->base.impl = key_pair_impl;
+    key_pair_impl->base.allocator = allocator;
+    aws_byte_buf_init_copy_from_cursor(&key_pair_impl->base.pub, allocator, key);
+
+    struct aws_der_decoder *decoder = aws_der_decoder_new(allocator, key);
+
+    if (!decoder) {
+        goto on_error;
+    }
+
+    struct s_rsa_public_key_pkcs1 public_key_data = {0};
```

Comment:
AWS_ZERO_STRUCT, as I recall this not working across several compilers?

Reply:
There is a couple places in der where we use the same annotation with structs and i've seen sporadic usage of this syntax throughout the codebase. dont mind switching to zero struct, but its something we should probably agree across team if this syntax is not valid

Reply:
It works if the first member in the struct can be set to 0. But like if the first member is another struct, that doesn't work.

AWS_ZERO_STRUCT is a bit more future-proof, in case the struct ever changes and it's first member can't be set 0 anymore

I don't super care either way